File: 第四阶段/4.3-文件读写/file_io.py
"""
Python 文件读写完整示例
对应文章：4.3 文件读写

C# 对比：
  Python open() + with     ≈  C# File.ReadAllText / using + StreamReader
  Python pathlib.Path      ≈  C# System.IO.Path + File
  Python io.StringIO       ≈  C# MemoryStream + StreamWriter
"""
import os
import io
import tempfile
import pathlib

# =============================================================
# 1. 文件打开模式 (open modes)
# =============================================================
# C# 对比：
#   "r"  ≈ File.OpenRead() / new StreamReader(path)
#   "w"  ≈ File.Create()   / new StreamWriter(path)
#   "a"  ≈ new StreamWriter(path, append: true)
#   "r+" ≈ new FileStream(path, FileMode.Open, FileAccess.ReadWrite)
#   "x"  ≈ new FileStream(path, FileMode.CreateNew)
#   "rb" ≈ File.ReadAllBytes(path)
#   "wb" ≈ File.WriteAllBytes(path, ...)
# =============================================================
print("=" * 60)
print("1. 文件打开模式 (open modes)")
print("=" * 60)

# 准备演示文件
test_dir = "_temp_fileio_demo"
os.makedirs(test_dir, exist_ok=True)
test_file = os.path.join(test_dir, "test.txt")

# --- "w" 模式：写入（覆盖） ---
# C# 等价：File.WriteAllText(path, content)
with open(test_file, "w", encoding="utf-8") as f:
    f.write("第一行\n")
    f.write("第二行\n")
    f.write("第三行\n")
print(f"['w' 模式] 写入完成: {test_file}")

# --- "r" 模式：只读 ---
# C# 等价：File.ReadAllText(path)
with open(test_file, "r", encoding="utf-8") as f:
    content = f.read()
    print(f"['r' 模式] 文件内容:\n{content}")

# --- "a" 模式：追加 ---
# C# 等价：new StreamWriter(path, append: true)
with open(test_file, "a", encoding="utf-8") as f:
    f.write("第四行(追加)\n")
print(f"['a' 模式] 追加完成")

with open(test_file, "r", encoding="utf-8") as f:
    print(f"['a' 模式] 追加后内容: {f.read()}")

# --- "x" 模式：排他创建（文件已存在则报错） ---
# C# 等价：new FileStream(path, FileMode.CreateNew)
new_file = os.path.join(test_dir, "new_only.txt")
try:
    with open(new_file, "x", encoding="utf-8") as f:
        f.write("新文件内容\n")
    print(f"['x' 模式] 创建成功: {new_file}")
    # 再次创建同一文件会报错
    with open(new_file, "x", encoding="utf-8") as f:
        f.write("这行不会执行")
except FileExistsError:
    print(f"['x' 模式] 文件已存在，创建失败 (FileExistsError)")

# --- "rb"/"wb" 二进制模式 ---
# C# 等价：File.ReadAllBytes / File.WriteAllBytes
binary_file = os.path.join(test_dir, "binary.bin")
data = bytes(range(256))  # 0x00 ~ 0xFF
with open(binary_file, "wb") as f:
    f.write(data)
print(f"['wb' 模式] 写入 {len(data)} 字节二进制数据")

with open(binary_file, "rb") as f:
    read_data = f.read()
print(f"['rb' 模式] 读取 {len(read_data)} 字节，内容匹配: {read_data == data}")

print()
print("=" * 60)
print("2. with 语句 — 自动资源管理")
print("=" * 60)

# -------------------------------------------------------
# C# 对比：
#   using (var reader = new StreamReader(path))
#   {
#       string line;
#       while ((line = reader.ReadLine()) != null)
#           Console.WriteLine(line);
#   }
#
# Python with ≈ C# using，都会在块结束时自动调用 close/Dispose
# -------------------------------------------------------
print("with 语句确保文件正确关闭，即使发生异常:")

with open(test_file, "r", encoding="utf-8") as f:
    for line in f:
        # f 本身是一个迭代器，逐行读取（内存友好）
        # C# 等价：reader.ReadLine() 逐行读取
        print(f"  {line.strip()}")

print()
print("=" * 60)
print("3. read / readlines / readline 方法")
print("=" * 60)

# -------------------------------------------------------
# C# 对比：
#   File.ReadAllText(path)       ≈  f.read()
#   File.ReadAllLines(path)      ≈  f.readlines()
#   reader.ReadLine()            ≈  f.readline()
# -------------------------------------------------------

# --- read() 读取全部内容 ---
with open(test_file, "r", encoding="utf-8") as f:
    full_content = f.read()  # 一次读入全部，适合小文件
print(f"read(): {full_content.strip()}")

# --- read(n) 读取指定字节数 ---
with open(test_file, "r", encoding="utf-8") as f:
    chunk = f.read(5)  # 读取前 5 个字符
print(f"read(5): '{chunk}'")

# --- readline() 逐行读取 ---
with open(test_file, "r", encoding="utf-8") as f:
    line1 = f.readline().strip()
    line2 = f.readline().strip()
print(f"readline(): '{line1}', '{line2}'")

# --- readlines() 读取所有行到列表 ---
with open(test_file, "r", encoding="utf-8") as f:
    lines = f.readlines()  # 返回 list[str]
print(f"readlines(): {lines}")

print()
print("=" * 60)
print("4. write / writelines 方法")
print("=" * 60)

# -------------------------------------------------------
# C# 对比：
#   File.WriteAllText(path, text)   ≈  f.write(text)
#   File.WriteAllLines(path, lines) ≈  f.writelines(lines)
# -------------------------------------------------------

output_file = os.path.join(test_dir, "output.txt")

# --- write() ---
# 注意：write() 不会自动添加换行符，需要手动添加
with open(output_file, "w", encoding="utf-8") as f:
    f.write("Hello, World!\n")
    f.write("你好，Python!\n")
    count = f.write("write() 返回写入的字符数\n")
    print(f"write() 返回: {count}")

# --- writelines() ---
# 注意：writelines() 也不会自动添加换行符
with open(output_file, "w", encoding="utf-8") as f:
    f.writelines(["line1\n", "line2\n", "line3\n"])

with open(output_file, "r", encoding="utf-8") as f:
    print(f"writelines() 写入结果: {f.read()}")

print()
print("=" * 60)
print("5. encoding 参数 — 指定编码")
print("=" * 60)

# -------------------------------------------------------
# C# 对比：
#   File.ReadAllText(path, Encoding.UTF8)
#   new StreamReader(path, Encoding.GetEncoding("gbk"))
# -------------------------------------------------------
encoding_file = os.path.join(test_dir, "encoding.txt")

# 写入不同编码
text = "编码测试: Hello 你好 [CELEBRATE]"

# UTF-8 (Python 默认)
with open(encoding_file, "w", encoding="utf-8") as f:
    f.write(text)

# 读取 UTF-8
with open(encoding_file, "r", encoding="utf-8") as f:
    print(f"UTF-8 读取: {f.read()}")

# GBK 编码
with open(encoding_file, "r", encoding="utf-8") as f:
    raw_bytes = f.read().encode("utf-8")
gbk_file = os.path.join(test_dir, "encoding_gbk.txt")
raw_bytes.decode("utf-8").encode("gbk", errors="replace")  # 不支持的字符会替换

with open(encoding_file, "rb") as f:
    binary = f.read()
print(f"文件大小 (UTF-8 编码): {len(binary)} 字节")

print()
print("=" * 60)
print("6. pathlib — 面向对象的路径操作 (推荐)")
print("=" * 60)

# -------------------------------------------------------
# C# 对比：
#   pathlib.Path  ≈  System.IO.Path + System.IO.File + System.IO.Directory
#   Python 3.4+ 引入，更 Pythonic 的文件操作方式
# -------------------------------------------------------
# 创建 Path 对象
p = pathlib.Path(test_dir) / "test.txt"
print(f"Path 对象: {p}")
print(f"  文件名: {p.name}")
print(f"  扩展名: {p.suffix}")
print(f"  父目录: {p.parent}")
print(f"  是否存在: {p.exists()}")
print(f"  文件大小: {p.stat().st_size} 字节")

# pathlib 读写文件（最简洁的方式）
content = p.read_text(encoding="utf-8")
print(f"  read_text(): {content.strip()}")

p.write_text("pathlib 写入的内容\n第二行\n", encoding="utf-8")
print(f"  write_text() 完成")

# 遍历目录
print(f"\n  目录 {test_dir} 下的文件:")
for item in pathlib.Path(test_dir).iterdir():
    print(f"    {'[DIR]' if item.is_dir() else '[FILE]'} {item.name}")

print()
print("=" * 60)
print("7. tempfile — 临时文件和临时目录")
print("=" * 60)

# -------------------------------------------------------
# C# 对比：
#   Path.GetTempFileName()     ≈  tempfile.NamedTemporaryFile()
#   Path.GetTempPath()         ≈  tempfile.gettempdir()
#   using var tmp = ...        ≈  with tempfile.NamedTemporaryFile() as tmp:
# -------------------------------------------------------
# 创建临时文件
with tempfile.NamedTemporaryFile(mode="w", suffix=".txt",
                                  delete=False, encoding="utf-8") as tmp:
    tmp.write("这是一个临时文件\n")
    tmp_name = tmp.name
    print(f"临时文件: {tmp_name}")

# 读取临时文件
with open(tmp_name, "r", encoding="utf-8") as f:
    print(f"读取: {f.read().strip()}")

# 创建临时目录
with tempfile.TemporaryDirectory() as tmp_dir:
    tmp_path = pathlib.Path(tmp_dir)
    tmp_file = tmp_path / "temp_data.txt"
    tmp_file.write_text("临时目录中的文件\n", encoding="utf-8")
    print(f"临时目录: {tmp_dir}")
    print(f"  目录内容: {list(tmp_path.iterdir())}")
# 离开 with 块后，临时目录和其中的文件自动删除

# 清理
os.unlink(tmp_name)

print()
print("=" * 60)
print("8. io.StringIO — 内存中的文件对象")
print("=" * 60)

# -------------------------------------------------------
# C# 对比：
#   StringIO ≈ MemoryStream + StreamWriter/StreamReader
#   用于在内存中模拟文件操作，常用于测试和字符串处理
# -------------------------------------------------------
# 创建 StringIO
buffer = io.StringIO()
buffer.write("第一行\n")
buffer.write("第二行\n")
buffer.write("第三行\n")

# seek 回到开头
buffer.seek(0)
content = buffer.read()
print(f"StringIO 内容:\n{content}")
buffer.close()

# 从字符串初始化
buffer2 = io.StringIO("Hello StringIO\n")
lines = buffer2.readlines()
print(f"从字符串初始化: {lines}")
buffer2.close()

print()
print("=" * 60)
print("9. 上下文管理器同时打开多个文件")
print("=" * 60)

# -------------------------------------------------------
# C# 对比：
#   using (var fin = new StreamReader(inPath))
#   using (var fout = new StreamWriter(outPath))
#   {
#       // ...
#   }
# -------------------------------------------------------
input_file = os.path.join(test_dir, "input.txt")
output_file = os.path.join(test_dir, "upper_output.txt")

with open(input_file, "w", encoding="utf-8") as f:
    f.write("hello world\n")
    f.write("python file io\n")

# Python 3.10+ 可以用括号包裹
with open(input_file, "r", encoding="utf-8") as fin, \
     open(output_file, "w", encoding="utf-8") as fout:
    for line in fin:
        fout.write(line.upper())

with open(output_file, "r", encoding="utf-8") as f:
    print(f"转换结果: {f.read().strip()}")

print()
print("=" * 60)
print("10. 文件操作速查表")
print("=" * 60)

# -------------------------------------------------------
# C# 对比：
#   File.Copy / File.Move / File.Delete / File.Exists
#   Directory.CreateDirectory / Directory.GetFiles
# -------------------------------------------------------
import shutil

# 文件操作
src_file = os.path.join(test_dir, "test.txt")
copy_file = os.path.join(test_dir, "test_copy.txt")
rename_file = os.path.join(test_dir, "test_renamed.txt")

shutil.copy(src_file, copy_file)
print(f"复制: {src_file} -> {copy_file}")

os.rename(copy_file, rename_file)
print(f"重命名: {copy_file} -> {rename_file}")

exists = os.path.exists(rename_file)
print(f"文件是否存在: {exists}")

size = os.path.getsize(rename_file)
print(f"文件大小: {size} 字节")

os.remove(rename_file)
print(f"删除: {rename_file}")

# 目录操作
new_dir = os.path.join(test_dir, "new_folder")
os.makedirs(new_dir, exist_ok=True)
print(f"创建目录: {new_dir}")

files = os.listdir(test_dir)
print(f"目录内容: {files}")

# 清理演示目录
shutil.rmtree(test_dir)
print(f"清理演示目录: {test_dir}")

print()
print("=" * 60)
print("11. 文件操作最佳实践总结")
print("=" * 60)
print("""
Python vs C# 文件操作对比:
  Python open() + with           ≈  C# File.ReadAllText / using StreamReader
  Python pathlib.Path             ≈  C# System.IO.Path + File
  Python io.StringIO              ≈  C# MemoryStream
  Python tempfile                 ≈  C# Path.GetTempFileName
  Python os.path / shutil         ≈  C# System.IO.File / Directory

最佳实践:
  1. 始终使用 with/using 确保资源释放
  2. 指定 encoding 参数，避免平台编码差异
  3. 小文件用 read()/write()，大文件用迭代逐行处理
  4. 路径操作优先使用 pathlib (Python) / Path (C#)
  5. 临时文件使用 tempfile 模块管理
""")

print("所有文件读写示例运行完毕！")

File: 第四阶段/4.3-文件读写/FileIO.cs
using System;
using System.IO;
using System.Text;

/// <summary>
/// C# 文件读写完整示例
/// 对应文章：4.3 文件读写
///
/// Python 对比：
///   C# File.ReadAllText / using StreamReader  ≈  Python open() + with
///   C# System.IO.Path + File                  ≈  Python pathlib.Path
///   C# MemoryStream                           ≈  Python io.StringIO
/// </summary>
class FileIO
{
    static void Main()
    {
        // =============================================================
        // 1. 文件读取方式
        // =============================================================
        // Python 对比：
        //   with open("file.txt", "r") as f:
        //       content = f.read()        # ≈ File.ReadAllText
        //       lines = f.readlines()     # ≈ File.ReadAllLines
        //       for line in f: ...        # ≈ reader.ReadLine() 循环
        // =============================================================
        Console.WriteLine("============================================================");
        Console.WriteLine("1. 文件读取方式");
        Console.WriteLine("============================================================");

        // 准备演示文件
        string testDir = "_temp_fileio_demo_cs";
        Directory.CreateDirectory(testDir);
        string testFile = Path.Combine(testDir, "test.txt");
        File.WriteAllText(testFile, "第一行\n第二行\n第三行\n", Encoding.UTF8);

        // --- File.ReadAllText: 一次读取全部 (Python: f.read()) ---
        string content = File.ReadAllText(testFile, Encoding.UTF8);
        Console.WriteLine($"File.ReadAllText():\n{content}");

        // --- File.ReadAllLines: 读取所有行到数组 (Python: f.readlines()) ---
        string[] lines = File.ReadAllLines(testFile, Encoding.UTF8);
        Console.WriteLine("File.ReadAllLines():");
        foreach (var line in lines)
            Console.WriteLine($"  {line}");

        // --- StreamReader 逐行读取 (Python: for line in f) ---
        Console.WriteLine("StreamReader 逐行读取:");
        using (var reader = new StreamReader(testFile, Encoding.UTF8))
        {
            string? line;
            while ((line = reader.ReadLine()) != null)
                Console.WriteLine($"  {line}");
        }
        // 离开 using 块，reader 自动 Dispose（等同 Python 的 with）

        // --- File.ReadAllBytes: 二进制读取 (Python: open("rb").read()) ---
        byte[] bytes = File.ReadAllBytes(testFile);
        Console.WriteLine($"File.ReadAllBytes(): {bytes.Length} 字节");

        // =============================================================
        // 2. 文件写入方式
        // =============================================================
        // Python 对比：
        //   with open("output.txt", "w") as f:
        //       f.write("text")          # ≈ File.WriteAllText / writer.Write
        //       f.writelines([...])      # ≈ File.WriteAllLines
        // =============================================================
        Console.WriteLine();
        Console.WriteLine("============================================================");
        Console.WriteLine("2. 文件写入方式");
        Console.WriteLine("============================================================");

        string outputFile = Path.Combine(testDir, "output.txt");

        // --- File.WriteAllText: 一次性写入 (Python: f.write(text)) ---
        File.WriteAllText(outputFile, "Hello, World!\n", Encoding.UTF8);
        Console.WriteLine("File.WriteAllText() 完成");

        // --- File.WriteAllLines: 写入多行 (Python: f.writelines([...])) ---
        File.WriteAllLines(outputFile, new[] { "line1", "line2", "line3" }, Encoding.UTF8);
        Console.WriteLine("File.WriteAllLines() 完成");

        // --- StreamWriter 写入 (Python: f.write()) ---
        string streamFile = Path.Combine(testDir, "stream_output.txt");
        using (var writer = new StreamWriter(streamFile, false, Encoding.UTF8))
        {
            writer.Write("StreamWriter 写入: ");
            writer.WriteLine("Hello");
            writer.WriteLine("StreamWriter 写入第二行");
        }
        Console.WriteLine("StreamWriter 写入完成");

        // --- 追加写入 (Python: open("file", "a")) ---
        using (var writer = new StreamWriter(outputFile, append: true, Encoding.UTF8))
        {
            writer.WriteLine("追加的内容");
        }
        Console.WriteLine("追加写入完成");

        // =============================================================
        // 3. using 语句 — 资源管理 (Python with 的等价物)
        // =============================================================
        // Python 对比：
        //   with open("file.txt") as f:
        //       content = f.read()
        //   # 离开 with 块自动关闭，不需要 finally
        // =============================================================
        Console.WriteLine();
        Console.WriteLine("============================================================");
        Console.WriteLine("3. using 语句 — 资源管理");
        Console.WriteLine("============================================================");

        // 传统 using 语句（等同 Python 的 with 语句）
        Console.WriteLine("传统 using 语句:");
        using (var reader = new StreamReader(testFile))
        using (var writer = new StreamWriter(Path.Combine(testDir, "copy.txt")))
        {
            string? line;
            int lineNum = 1;
            while ((line = reader.ReadLine()) != null)
            {
                Console.WriteLine($"  {lineNum++}: {line}");
                writer.WriteLine(line);
            }
        }

        // C# 8+ using 声明（更简洁）
        Console.WriteLine("C# 8+ using 声明:");
        using var reader2 = new StreamReader(testFile);
        Console.WriteLine($"  第一行: {reader2.ReadLine()}");

        // =============================================================
        // 4. 文件模式和 FileOptions
        // =============================================================
        // Python 对比：
        //   open("file", "r")   → FileMode.Open
        //   open("file", "w")   → FileMode.Create (覆盖)
        //   open("file", "a")   → FileMode.Append
        //   open("file", "x")   → FileMode.CreateNew (不存在才创建)
        //   open("file", "r+")  → FileMode.Open, FileAccess.ReadWrite
        // =============================================================
        Console.WriteLine();
        Console.WriteLine("============================================================");
        Console.WriteLine("4. 文件模式和 FileOptions");
        Console.WriteLine("============================================================");

        // FileMode.CreateNew (Python: "x" 模式)
        string newOnlyFile = Path.Combine(testDir, "new_only.txt");
        try
        {
            using var fs = new FileStream(newOnlyFile, FileMode.CreateNew);
            byte[] data = Encoding.UTF8.GetBytes("新文件内容");
            fs.Write(data, 0, data.Length);
            Console.WriteLine($"CreateNew: 创建成功 {newOnlyFile}");

            // 再次创建会报错
            using var fs2 = new FileStream(newOnlyFile, FileMode.CreateNew);
        }
        catch (IOException ex)
        {
            Console.WriteLine($"CreateNew: 文件已存在 — {ex.Message}");
        }

        // FileMode.OpenOrCreate (Python: open("file", "a") 的变体)
        string openOrCreate = Path.Combine(testDir, "open_or_create.txt");
        using (var fs = new FileStream(openOrCreate, FileMode.OpenOrCreate))
        {
            byte[] data = Encoding.UTF8.GetBytes("OpenOrCreate 写入");
            fs.Write(data, 0, data.Length);
        }
        Console.WriteLine("OpenOrCreate 完成");

        // =============================================================
        // 5. 编码 — Encoding 参数
        // =============================================================
        // Python 对比：
        //   open("file", "r", encoding="utf-8")
        //   open("file", "r", encoding="gbk")
        // =============================================================
        Console.WriteLine();
        Console.WriteLine("============================================================");
        Console.WriteLine("5. 编码 — Encoding 参数");
        Console.WriteLine("============================================================");

        string encodingFile = Path.Combine(testDir, "encoding.txt");
        string text = "编码测试: Hello 你好";

        // UTF-8 编码
        File.WriteAllText(encodingFile, text, Encoding.UTF8);
        string readBack = File.ReadAllText(encodingFile, Encoding.UTF8);
        Console.WriteLine($"UTF-8 读取: {readBack}");

        // 编码信息
        byte[] utf8Bytes = Encoding.UTF8.GetBytes(text);
        byte[] gbkBytes = Encoding.GetEncoding("gbk").GetBytes(text);
        Console.WriteLine($"UTF-8 字节数: {utf8Bytes.Length}");
        Console.WriteLine($"GBK 字节数: {gbkBytes.Length}");

        // =============================================================
        // 6. Path 类 — 路径操作 (Python pathlib 的等价物)
        // =============================================================
        // Python 对比：
        //   p = pathlib.Path("dir") / "file.txt"
        //   p.name, p.suffix, p.parent, p.stem
        // =============================================================
        Console.WriteLine();
        Console.WriteLine("============================================================");
        Console.WriteLine("6. Path 类 — 路径操作");
        Console.WriteLine("============================================================");

        string fullPath = Path.GetFullPath(testFile);
        Console.WriteLine($"完整路径: {fullPath}");
        Console.WriteLine($"  文件名 (Name): {Path.GetFileName(fullPath)}");
        Console.WriteLine($"  无扩展名 (Stem): {Path.GetFileNameWithoutExtension(fullPath)}");
        Console.WriteLine($"  扩展名 (Extension): {Path.GetExtension(fullPath)}");
        Console.WriteLine($"  父目录 (DirectoryName): {Path.GetDirectoryName(fullPath)}");
        Console.WriteLine($"  是否绝对路径: {Path.IsPathRooted(fullPath)}");

        // 路径组合 (Python: p1 / p2)
        string combined = Path.Combine("C:", "Users", "test", "file.txt");
        Console.WriteLine($"  Path.Combine: {combined}");

        // 临时路径
        Console.WriteLine($"  临时目录: {Path.GetTempPath()}");
        Console.WriteLine($"  临时文件: {Path.GetTempFileName()}");

        // =============================================================
        // 7. MemoryStream — 内存中的文件 (Python io.StringIO 的等价物)
        // =============================================================
        // Python 对比：
        //   import io
        //   buffer = io.StringIO()
        //   buffer.write("text")
        //   buffer.seek(0)
        //   content = buffer.read()
        // =============================================================
        Console.WriteLine();
        Console.WriteLine("============================================================");
        Console.WriteLine("7. MemoryStream — 内存中的文件");
        Console.WriteLine("============================================================");

        // 写入 MemoryStream
        using var memoryStream = new MemoryStream();
        using (var writer = new StreamWriter(memoryStream, Encoding.UTF8, leaveOpen: true))
        {
            writer.WriteLine("第一行");
            writer.WriteLine("第二行");
            writer.Flush();
        }

        // 读取 MemoryStream
        memoryStream.Seek(0, SeekOrigin.Begin);
        using var reader3 = new StreamReader(memoryStream);
        string memoryContent = reader3.ReadToEnd();
        Console.WriteLine($"MemoryStream 内容:\n{memoryContent}");

        // =============================================================
        // 8. 临时文件和临时目录
        // =============================================================
        // Python 对比：
        //   import tempfile
        //   with tempfile.NamedTemporaryFile() as tmp:
        //       tmp.write(b"data")
        //   with tempfile.TemporaryDirectory() as tmp_dir:
        //       ...
        // =============================================================
        Console.WriteLine();
        Console.WriteLine("============================================================");
        Console.WriteLine("8. 临时文件和临时目录");
        Console.WriteLine("============================================================");

        // 临时文件
        string tempFile = Path.GetTempFileName();
        File.WriteAllText(tempFile, "临时文件内容", Encoding.UTF8);
        Console.WriteLine($"临时文件: {tempFile}");
        Console.WriteLine($"  内容: {File.ReadAllText(tempFile)}");

        // 临时目录
        string tempDir = Path.Combine(Path.GetTempPath(), $"demo_{Guid.NewGuid():N}");
        Directory.CreateDirectory(tempDir);
        File.WriteAllText(Path.Combine(tempDir, "temp.txt"), "临时数据", Encoding.UTF8);
        Console.WriteLine($"临时目录: {tempDir}");
        Console.WriteLine($"  文件: {string.Join(", ", Directory.GetFiles(tempDir).Select(Path.GetFileName))}");

        // 清理
        File.Delete(tempFile);
        Directory.Delete(tempDir, true);
        Console.WriteLine("临时文件和目录已清理");

        // =============================================================
        // 9. 文件异步读写 (Python asyncio 的对比)
        // =============================================================
        Console.WriteLine();
        Console.WriteLine("============================================================");
        Console.WriteLine("9. 文件异步读写");
        Console.WriteLine("============================================================");
        Console.WriteLine("C# 支持原生异步文件 I/O，Python 也可以用 aiofiles 库");
        Console.WriteLine("C# 优势: File.ReadAllTextAsync / File.WriteAllTextAsync");
        Console.WriteLine("Python: 默认是同步的，需要 aiofiles 或 asyncio.to_thread");

        // 注意：async Main 需要 C# 7.1+
        // string asyncContent = await File.ReadAllTextAsync(testFile);
        // await File.WriteAllTextAsync("async.txt", "Hello");
        Console.WriteLine("  (异步方法示例，需要 async Main 声明)");

        // =============================================================
        // 10. 文件操作速查表
        // =============================================================
        Console.WriteLine();
        Console.WriteLine("============================================================");
        Console.WriteLine("10. 文件操作速查表");
        Console.WriteLine("============================================================");

        string copyFile = Path.Combine(testDir, "test_copy.txt");
        string renameFile = Path.Combine(testDir, "test_renamed.txt");

        // 复制
        File.Copy(testFile, copyFile);
        Console.WriteLine($"复制: {testFile} -> {copyFile}");

        // 重命名
        File.Move(copyFile, renameFile);
        Console.WriteLine($"重命名: {copyFile} -> {renameFile}");

        // 检查存在
        bool exists = File.Exists(renameFile);
        Console.WriteLine($"文件是否存在: {exists}");

        // 获取大小
        var fileInfo = new FileInfo(renameFile);
        Console.WriteLine($"文件大小: {fileInfo.Length} 字节");

        // 删除
        File.Delete(renameFile);
        Console.WriteLine($"删除: {renameFile}");

        // 目录操作
        string newDir = Path.Combine(testDir, "new_folder");
        Directory.CreateDirectory(newDir);
        Console.WriteLine($"创建目录: {newDir}");

        string[] files = Directory.GetFiles(testDir);
        Console.WriteLine($"目录内容: {string.Join(", ", files.Select(Path.GetFileName))}");

        // 清理
        Directory.Delete(testDir, true);
        Console.WriteLine($"清理演示目录: {testDir}");

        // =============================================================
        // 11. 文件操作最佳实践总结
        // =============================================================
        Console.WriteLine();
        Console.WriteLine("============================================================");
        Console.WriteLine("11. 文件操作最佳实践总结");
        Console.WriteLine("============================================================");
        Console.WriteLine();
        Console.WriteLine("C# vs Python 文件操作对比:");
        Console.WriteLine("  C# File.ReadAllText           ≈  Python open().read()");
        Console.WriteLine("  C# using StreamReader          ≈  Python with open() as f");
        Console.WriteLine("  C# Path                       ≈  Python pathlib.Path");
        Console.WriteLine("  C# MemoryStream               ≈  Python io.StringIO");
        Console.WriteLine("  C# Path.GetTempFileName       ≈  Python tempfile.NamedTemporaryFile");
        Console.WriteLine("  C# File.Copy/Move/Delete      ≈  Python shutil.copy/move + os.remove");
        Console.WriteLine("  C# Directory.GetFiles         ≈  Python os.listdir / pathlib.iterdir");
        Console.WriteLine();
        Console.WriteLine("最佳实践:");
        Console.WriteLine("  1. 始终使用 using/with 确保资源释放");
        Console.WriteLine("  2. 指定 Encoding 参数，避免平台编码差异");
        Console.WriteLine("  3. 小文件用 ReadAllText/WriteAllText，大文件用 StreamReader");
        Console.WriteLine("  4. 路径操作优先使用 Path (C#) / pathlib (Python)");
        Console.WriteLine("  5. 临时文件使用 Path.GetTempFileName / tempfile 模块");

        Console.WriteLine();
        Console.WriteLine("所有文件读写示例运行完毕！");
    }
}
